src/validation.cpp

```diff
@@ -1593,6 +1593,13 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex& block_index, const Co
     {
         // Enforce WITNESS rules whenever P2SH is in effect
         flags |= SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS;
+        // Enforce Taproot (BIP340-BIP342) whenever WITNESS is in effect,
+        // unless except
+        if (consensusparams.TaprootException.IsNull() ||                         // no Taproot exception on this chain
+            *Assert(block_index.phashBlock) != consensusparams.TaprootException) // this block is not the historical exception
```

Comment:
Arguably it would also be correct to write this as:

```c++
const auto& h = *Assert(block_index.phashBlock);
if (h.IsNull() && h != BIP16Exception && h != TaprootException) {
    flags |= SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS | SCRIPT_VERIFY_TAPROOT;
}
```

since the taproot exception block has already been validated to comply with p2sh and segwit tx rules (and the inclusion of witness data is handled via BLOCK_OPT_WITNESS elsewhere)

Reply:
It seems there is a typo in your idea. `h.IsNull()` should never evaluate to true. Maybe you meant to write

```cpp
const auto& h = *Assert(block_index.phashBlock);
if (BIP16Exception.IsNull() || TaprootException.IsNull() || (h != BIP16Exception && h != TaprootException)) {
    flags |= SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS | SCRIPT_VERIFY_TAPROOT;
}
```

Though, this doesn't work, as I expect it to fail to sync testnet3.

As I mentioned earlier, just dropping the IsNull check seems simpler (https://github.com/bitcoin/bitcoin/pull/23536#discussion_r759927632).

So your code would be 

```cpp
const auto& h = *Assert(block_index.phashBlock);
if (h != BIP16Exception && h != TaprootException) {
    flags |= SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS | SCRIPT_VERIFY_TAPROOT;
}
```

which should work (Haven't synced any chain with it yet). Happy to switch if reviewers think it is worth it.

Reply:
I don't think it's worth including in this PR -- if it's worth including at all, doing so in an eventual followup is fine.

If we're sure `h.IsNull()` is never true, then `BIP16Exception.IsNull()` implies `h != BIP16Exception` so you don't need the `IsNull` test at all. I thought it was supposed to be able to be null if you were generating a new block and trying to work out which txs you were allowed to include.
The way I'm thinking of this is:

 * features that will be used by many wallets (segwit, taproot, p2sh) should be "always" enforced if possible; this reduces the risk of bugs that might be triggered by an attacker against a node while it's doing IBD
 
 * there have to be some exceptions unfortunately, but the exceptions aren't the interesting part at all, so should be emphasising the enforcement path more than the exception logic
 
  * since the block hash already commits to everything that goes into proving the validity, actually running the checks is only useful to validate the code hasn't changed it's behaviour, not that the block is valid or not. that is, if we ever report that the exception blocks are invalid, that's a bug in the code, not a problem with the block.

Maybe a cleaner approach would be:

```c++
// add to Consensus::Params or CChainState
// (lists the rules that *don't* apply to exception blocks)
const unordered_map<uint256, uint32_t, BlockHasher> script_verify_exception_blocks = {
    { BIP16Exception, SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS | SCRIPT_VERIFY_TAPROOT },
    { TaprootException, SCRIPT_VERIFY_TAPROOT },
};

// validation:
flags |= SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS | SCRIPT_VERIFY_TAPROOT;

const auto& i = script_verify_exception_blocks.find(*Assert(block_index.phashBlock))
if (i != script_verify_exception_blocks.end()) {
    // block is special case exception to some rules
    flags &= ~(i->second);
}
```

Then for chains where there's no exceptions, the map is just empty.

Reply:
> If we're sure h.IsNull() is never true

Yes, I think we can be pretty certain about this. Both call sites assume the hash of the passed block is not null:

E.g.

```
    assert(*pindex->phashBlock == block.GetHash());
```

The `IsNull` was added based on this comment: https://github.com/bitcoin/bitcoin/pull/11739#discussion_r160814301

Reply:
Not sure if I like your suggestion to split "flags math" into two different files. This would make reading the code harder. Based on your earlier suggestion, just treating a set of block hashes as "assumed valid" with regard to script verification flags seems cleaner:

```cpp

const std::set<uint256> script_verify_exception_blocks;

const auto& it_ex{script_verify_exception_blocks.find(*Assert(block_index.phashBlock))};
if (it_ex != script_verify_exception_blocks.end()) {
    flags |= SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS | SCRIPT_VERIFY_TAPROOT;
}

Reply:
Putting the map in `CChainState` rather than `Consensus::Params` would allow keeping it validation rather than splitting it into validation/chainparams fwiw. Agree that your suggestion also works fine.

Reply:
If you're using a `set` rather than a `map`, you can just say `script_verify_exception_blocks.count(...) != 0` and avoid the iterator and the `.end()` check.

Reply:
Ok, done in https://github.com/bitcoin/bitcoin/pull/23536#issuecomment-997790605. Happy to revert to previous ACKed commit if needed.